urllib3/_collections.py

```diff
@@ -143,7 +141,10 @@ class HTTPHeaderDict(dict):
     def __init__(self, headers=None, **kwargs):
         dict.__init__(self)
         if headers is not None:
-            self.extend(headers)
+            if isinstance(headers, type(self)):
```

Comment:
Here I'd use HTTPHeaderDict explicitly since we probably want to keep this behavior on inherited classes and type(self) will return the subclass instead of the parent.